python/setup.py

```diff
@@ -82,18 +91,81 @@
     "tune": ["tabulate"],
 }
 
+def is_native_windows_or_msys():  # Does NOT return true for WSL, which is considered "Linux"
+    return sys.platform == 'msys' or sys.platform == 'win32'
+
+def is_invalid_windows_platform():
+    return sys.platform == 'msys' or sys.platform == 'win32' and sys.version_string and 'GCC' in sys.version_string
+
+def makedirs(path, exist_ok):
+    # WARNING: path components to create should NOT include ".." components
+    # Documentation: https://docs.python.org/3/library/os.html#os.makedirs
+    if sys.version_info[:2] >= (3, 2):
+        os.makedirs(path, exist_ok=exist_ok)
+    else:
+        try:
+            os.makedirs(path)
+        except OSError as ex:
+            error_due_to_existing_dir = exist_ok and ex.errno == errno.EEXIST and os.path.isdir(path)
+            if not error_due_to_existing_dir:
+                raise
+
+def download(url):
+    return urllib_request.urlopen(url).read()
+
+def bazel_build(root_dir, build_ext=None):  # Keep this as a global function to make life easier during development
+    if is_native_windows_or_msys():
+        BUILD_SH = os.getenv('BUILD_SH')
+        SYSTEMROOT = os.getenv('SYSTEMROOT')
+        if (not BUILD_SH) and SYSTEMROOT and os.path.isfile(os.path.join(SYSTEMROOT, 'System32', 'bash.exe')):
+            msg = " ".join([
+                "You appear to have Bash from WSL, which Bazel is not compatible with.",
+                "To avoid potential problems, please explicitly set the %r environment variable for Bazel."
+            ]) % ('BUILD_SH',)
+            raise ValueError(msg)
+
+    assert os.path.isabs(root_dir), "need current working directory as an absolute path"
```

Comment:
`assert` statements should only be used for internal invariant checking. Raise a proper exception if this is validation